File: iosApp/iosApp/ComponentViews/DirectionLabel.swift
//
//  DirectionLabel.swift
//  iosApp
//
//  Created by Simon, Emma on 6/24/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Shared
import SwiftUI

struct DirectionLabel: View {
    @ObserveInjection var inject
    let direction: Direction
    let showDestination: Bool
    let pillDecoration: PredictionRowView.PillDecoration
    let isTitle: Bool
    let routeNamePrefix: String?

    init(
        direction: Direction,
        showDestination: Bool = true,
        pillDecoration: PredictionRowView.PillDecoration = .none,
        isTitle: Bool = false,
        routeNamePrefix: String? = nil,
    ) {
        self.direction = direction
        self.showDestination = showDestination
        self.pillDecoration = pillDecoration
        self.isTitle = isTitle
        self.routeNamePrefix = routeNamePrefix
    }

    private static let localizedDirectionNames: [String: String] = [
        "North": NSLocalizedString("Northbound", comment: "A route direction label"),
        "South": NSLocalizedString("Southbound", comment: "A route direction label"),
        "East": NSLocalizedString("Eastbound", comment: "A route direction label"),
        "West": NSLocalizedString("Westbound", comment: "A route direction label"),
        "Inbound": NSLocalizedString("Inbound", comment: "A route direction label"),
        "Outbound": NSLocalizedString("Outbound", comment: "A route direction label"),
    ]

    static func directionNameFormatted(_ direction: Direction) -> String {
        localizedDirectionNames[direction.name ?? ""] ??
            NSLocalizedString("Heading", comment: "A route direction label")
    }

    @ViewBuilder
    func directionTo(_ direction: Direction) -> some View {
        let routeNamePrefix = if let routeNamePrefix { "\(routeNamePrefix) " } else { "" }
        let directionTo = String(format: NSLocalizedString("%@ to", comment: """
        Label the direction a list of arrivals is for.
        Possible values include Northbound, Southbound, Inbound, Outbound, Eastbound, Westbound.
        For example, "[Northbound] to [Alewife]"
        """), Self.directionNameFormatted(direction))
        Text(routeNamePrefix + directionTo)
            .font(Typography.footnote)
            .textCase(.none)
            .multilineTextAlignment(.leading)
    }

    @ViewBuilder
    func destinationLabel(_ destination: String) -> some View {
        HStack(alignment: .center, spacing: 8) {
            if case let .onDirectionDestination(route: route) = pillDecoration {
                RoutePill(route: route, type: .flex)
            }
            Text(destination)
                .font(isTitle ? Typography.title3Bold : Typography.bodySemibold)
                .multilineTextAlignment(.leading)
                .textCase(.none)
        }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: isTitle ? 2 : 6) {
            if !showDestination {
                directionTo(direction)
            } else if let destination = direction.destination {
                directionTo(direction)
                destinationLabel(destination)
            } else {
                destinationLabel(DirectionLabel.directionNameFormatted(direction))
            }
        }.accessibilityElement(children: .combine)
            .enableInjection()
    }
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/component/DirectionLabel.kt
package com.mbta.tid.mbta_app.android.component

import androidx.annotation.StringRes
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.util.Typography
import com.mbta.tid.mbta_app.android.util.modifiers.placeholderIfLoading
import com.mbta.tid.mbta_app.model.Direction

private val localizedDirectionNames: Map<String, Int> =
    mapOf(
        "North" to R.string.northbound,
        "South" to R.string.southbound,
        "East" to R.string.eastbound,
        "West" to R.string.westbound,
        "Inbound" to R.string.inbound,
        "Outbound" to R.string.outbound,
    )

@StringRes
fun directionNameFormatted(direction: Direction) =
    localizedDirectionNames[direction.name] ?: R.string.heading

@Composable
fun DirectionTo(
    direction: Direction,
    textColor: Color,
    onlyServingOppositeDirection: Boolean = false,
) {
    Text(
        stringResource(
            if (onlyServingOppositeDirection) R.string.only_direction_to else R.string.directionTo,
            stringResource(directionNameFormatted(direction)),
        ),
        color = textColor,
        modifier = Modifier.placeholderIfLoading(),
        style = if (onlyServingOppositeDirection) Typography.footnoteItalic else Typography.footnote,
    )
}

@Composable
fun DestinationLabel(direction: Direction, textColor: Color, pillDecoration: PillDecoration?) {
    DestinationLabel(stringResource(directionNameFormatted(direction)), textColor, pillDecoration)
}

@Composable
fun DestinationLabel(destination: String, textColor: Color, pillDecoration: PillDecoration?) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (pillDecoration is PillDecoration.OnDirectionDestination) {
            RoutePill(pillDecoration.route, line = null, RoutePillType.Flex)
        }
        Text(
            destination,
            color = textColor,
            modifier = Modifier.placeholderIfLoading(),
            style = Typography.bodySemibold,
        )
    }
}

@Composable
fun DirectionLabel(
    direction: Direction,
    modifier: Modifier = Modifier,
    textColor: Color = LocalContentColor.current,
    showDestination: Boolean = true,
    pillDecoration: PillDecoration? = null,
    onlyServingOppositeDirection: Boolean = false,
) {
    val destination = direction.destination
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(6.dp)) {
        if (!showDestination) {
            DirectionTo(direction, textColor, onlyServingOppositeDirection)
        } else if (destination != null) {
            DirectionTo(direction, textColor, onlyServingOppositeDirection)
            DestinationLabel(destination, textColor, pillDecoration)
        } else {
            DestinationLabel(direction, textColor, pillDecoration)
        }
    }
}
